### PR TITLE
cdc: don't emit resolved timestamps before scan

### PR DESCRIPTION
**Pull request by hicqu**
Signed-off-by: qupeng <qupeng@pingcap.com>

Issue Number: Close https://github.com/pingcap/tiflow/issues/4782

CDC shouldn't emit resolved timestamps before incremental scan finished. https://github.com/tikv/tikv/pull/11098 tried to fix the problem but the implementation is wrong. This patch gives a correct solution.

```commit-message
cdc: don't emit resolved timestamps before scan
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test

Side effects

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/cdc/src/endpoint.rs [hunk 1]
```diff
@@ -1253,23 +1257,24 @@ impl<E: KvEngine> Initializer<E> {
...
+                "conn_id" => ?conn_id);
+            Err(box_err!("scan canceled"))
+        };
 
         while !done {
             // When downstream_state is Stopped, it means the corresponding
             // delegate is stopped. The initialization can be safely canceled.
             if self.downstream_state.load() == DownstreamState::Stopped {
-                on_cancel!();
+                on_cancel()?;
```
Comment (overvenus) [inline components/cdc/src/endpoint.rs:1277]: What about `return on_cancel();`? So we know for sure that it is returned here.

File: components/cdc/tests/failpoints/test_endpoint.rs [hunk 2]
```diff
@@ -251,52 +251,33 @@ fn test_no_resolved_ts_before_downstream_initialized() {
...
-        let (mut req_tx, event_feed, receive_event) =
-            new_event_feed(suite.get_region_cdc_client(region.id));
+        let (mut req_tx, event_feed, _) = new_event_feed(suite.get_region_cdc_client(region.id));
         let req = suite.new_changedata_request(region.id);
         block_on(req_tx.send((req, WriteFlags::default()))).unwrap();
         req_txs.push(req_tx);
         event_feeds.push(event_feed);
-        receive_events.push(receive_event);
         // Sleep a while to wait the capture has been initialized.
         thread::sleep(Duration::from_secs(1));
```
Comment (overvenus) [inline components/cdc/tests/failpoints/test_endpoint.rs:268]: After sleep could you wait a resolved ts event? So we known it is initialized.

File: components/cdc/src/delegate.rs [hunk 3]
```diff
@@ -61,6 +66,40 @@ impl Default for DownstreamState {
     }
 }
 
+/// Shold only be called when it's uninitialized or stopped. Return false if it's stopped.
+pub fn on_init_downstream(s: &AtomicCell<DownstreamState>) -> bool {
```
Comment (hi-rustin) [inline components/cdc/src/delegate.rs:70]: I think the visibility of these methods could be smaller, for example they should be pub(crate). Using smaller visibility would help us to find dead code.

**Review by hi-rustin — APPROVED** [review]

**Review by overvenus — APPROVED** [review]
LGTM

**Pull request merged**